src/rdf4cpp/rdf/util/BigDecimal.hpp

```diff
@@ -0,0 +1,859 @@
+#ifndef RDF4CPP_BIGDECIMAL_H
+#define RDF4CPP_BIGDECIMAL_H
+
+#include <functional>
+#include <sstream>
+#include <string>
+#include <string_view>
+
+#include <boost/multiprecision/cpp_int.hpp>
+#include <nonstd/expected.hpp>
+
+namespace rdf4cpp::rdf::util {
+enum class RoundingMode {
+    Floor,
+    Ceil,
+    Round,
+};
+
+enum class Sign {
+    Positive = 0,
+    Negative = 1,
+};
+[[nodiscard]] constexpr Sign operator^(Sign a, Sign b) noexcept {
+    using T = std::underlying_type_t<Sign>;
+    return static_cast<Sign>(static_cast<T>(static_cast<T>(a) ^ static_cast<T>(b)));
+}
+[[nodiscard]] constexpr Sign operator!(Sign a) noexcept {
+    using T = std::underlying_type_t<Sign>;
+    return static_cast<Sign>(static_cast<T>(!static_cast<T>(a)));
+}
+
+enum class DecimalError {
+    Overflow,
+    NotDefined,  // aka NotANumber
+};
+
+template<typename T>
+concept BigDecimalBaseType = (std::unsigned_integral<T> && sizeof(T) >= sizeof(int32_t)) || std::same_as<T, boost::multiprecision::cpp_int>;
+
+template<BigDecimalBaseType UnscaledValue_t = boost::multiprecision::cpp_int, BigDecimalBaseType Exponent_t = uint32_t>
+class BigDecimal {
+    UnscaledValue_t unscaled_value;
+    Exponent_t exponent;
+    Sign sign = Sign::Positive;
```

Comment:
to use unsigned integers in case of not using cpp_int

Reply:
Why can't you just use signed ones?

Reply:
To avoid the asymetry between positive and negative numbers.

Reply:
Is avoiding that required for anything?

Reply:
Removed the Sign enum completely and changed the UnscaledValue_t to signed.
Drawback: Subtraction is now its own function and no longer just an add with unary minus applied to one of the operands. And unary minus/abs can now theoretically overflow. (Both because `-std::numeric_limits<int>::min()` is undefined)